plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitManager.java

```diff
@@ -0,0 +1,195 @@
+                        jdbcTableHandle.getRequiredNamedRelation().getRemoteTableName()));
+
+        if (!isUnloadSupported(((JdbcTableHandle) table), columns)) {
+            log.debug("Unload query contains unsupported characters. Falling back to using JDBC");
+            return fallbackSplitSource;
+        }
+
+        Connection connection;
+        PreparedStatement statement;
+        try {
+            connection = jdbcClient.getConnection(session);
+            String redshiftSelectSql = getRedshiftSelectSql(session, connection, jdbcTableHandle, columns);
+            if (redshiftSelectSql.contains("\\b")) {
+                log.debug("Unload query contains unsupported characters. Falling back to using JDBC");
+                return fallbackSplitSource;
+            }
+            statement = buildUnloadSql(session, connection, columns, redshiftSelectSql);
+        }
+        catch (SQLException e) {
+            if (e instanceof RedshiftException) {
+                log.debug("Could not build unload query. Falling back to using JDBC");
+                return fallbackSplitSource;
+            }
+            throw new RuntimeException(e);
+        }
+        return new RedshiftUnloadSplitSource(executor, connection, statement, fallbackSplitSource);
+    }
+
+    private String getRedshiftSelectSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = jdbcClient.prepareQuery(session, table, Optional.empty(), columns, ImmutableMap.of());
+        PreparedStatement openTelemetryPreparedStatement = queryBuilder.prepareStatement(jdbcClient, session, connection, preparedQuery, Optional.of(columns.size()));
+        RedshiftPreparedStatement redshiftPreparedStatement = openTelemetryPreparedStatement.unwrap(RedshiftPreparedStatement.class);
+        String selectQuerySql = redshiftPreparedStatement.toString();
+        return queryModifier.apply(session, selectQuerySql); // TODO is this required?
+    }
+
+    private PreparedStatement buildUnloadSql(ConnectorSession session, Connection connection, List<JdbcColumnHandle> columns, String redshiftSelectSql)
+            throws SQLException
+    {
+        String unloadSql = "UNLOAD ('%s') TO '%s' %s FORMAT PARQUET".formatted(
+                formatStringLiteral(redshiftSelectSql), // TODO does it require any other escaping
+                unloadLocation + session.getQueryId() + "-" + UUID.randomUUID() + "/",
+                unloadAuthorization);
+        return queryBuilder.prepareStatement(jdbcClient, session, connection, new PreparedQuery(unloadSql, List.of()), Optional.of(columns.size()));
+    }
+
+    private static String formatStringLiteral(String x)
+    {
+        return x.replace("'", "''");
+    }
+
+    private static boolean isUnloadSupported(JdbcTableHandle table, List<JdbcColumnHandle> columns)
```

Comment:
spark-redshift fails the queries in case of any failures but it doesn't seems to have special handling for certain situation. Most of the limitations are with the UNLOAD except with the decimal type which has to do with the way we get the Redshift select query from the drivers, spark-redshift doesn't have decimal type limit as it handcraft the query without using redshift drivers. I haven't been able to run it locally but will try to see what happens in such cases.